src/python/pants/core/goals/generate_lockfiles.py

```diff
@@ -378,27 +387,41 @@ async def generate_lockfiles_goal(
         set(generate_lockfiles_subsystem.resolve),
     )
 
+    # This is the "planning" phase of lockfile generation. Currently this is all done in the local
+    # environment, since there's not currently a clear mechanism to proscribe an environment.
     all_specified_user_requests = await MultiGet(
-        Get(UserGenerateLockfiles, RequestedUserResolveNames, resolve_names)
+        Get(
+            UserGenerateLockfiles,
+            {resolve_names: RequestedUserResolveNames, local_environment.val: EnvironmentName},
+        )
         for resolve_names in requested_user_resolve_names
     )
     specified_tool_requests = await MultiGet(
-        Get(WrappedGenerateLockfile, GenerateToolLockfileSentinel, sentinel())
+        Get(
+            WrappedGenerateLockfile,
+            {sentinel(): GenerateToolLockfileSentinel, local_environment.val: EnvironmentName},
+        )
         for sentinel in requested_tool_sentinels
     )
     applicable_tool_requests = filter_tool_lockfile_requests(
         specified_tool_requests,
         resolve_specified=bool(generate_lockfiles_subsystem.resolve),
     )
 
+    # Currently, since resolves specify a single filename for output, we pick a resonable
+    # environment to execute the request in. Currently we warn if multiple environments are
+    # specified.
+    all_requests = itertools.chain(*all_specified_user_requests, applicable_tool_requests)
+    preferred_envs = (_preferred_environment(req, local_environment.val) for req in all_requests)
```

Comment:
I assume that this is working, but it's a little bit unnerving to me, because I expect that both of these are iterators, and iterators can be consumed as side-effects. It's not clear that the `zip` below would see `all_requests` "from the beginning".

Maybe call `_preferred_environment` inside of the `MultiGet` rather than zipping it on?

Reply:
it wasn't  :)

good catch, did the suggested rewrite